torch/fx/graph.py

```diff
@@ -30,6 +60,138 @@ def _snake_case(s: str) -> str:
         prev_lower = c.islower()
     return ''.join(chars)
 
+
+def _is_from_torch(obj: Any) -> bool:
+    module_name = getattr(obj, '__module__', None)
+    if module_name is not None:
+        base_module = module_name.partition('.')[0]
+        return base_module == 'torch'
+
+    name = getattr(obj, '__name__', None)
+    # exclude torch because torch.torch.torch.torch works. idk mang
+    if name is not None and name != 'torch':
+        for guess in [torch, torch.nn.functional]:
+            if getattr(guess, name, None) is obj:
+                return True
+
+    return False
+
+
+class _NameContext:
+    def __init__(self):
+        self._obj_to_name: Dict[Any, str] = {}
+        self._used_names: Dict[str, int] = {}
+
+    def create_name(self, candidate: str, obj: Optional[Any] = None) -> str:
+        if obj is not None and obj in self._obj_to_name:
+            return self._obj_to_name[obj]
+
+        # delete all characters that are illegal in a Python identifier
+        candidate = re.sub('[^0-9a-zA-Z_]+', '_', candidate)
+        if candidate[0].isdigit():
+            candidate = f'_{candidate}'
+
+        while self._is_used(candidate) or self._is_illegal_name(candidate, obj):
+            match = re.match(r"(.*)_(\d+)$", candidate)
+            if match is None:
+                candidate = candidate + '_1'
+            else:
+                base, num = match.group(1, 2)
+                candidate = f'{base}_{int(num) + 1}'
+
+        self._use_name(candidate)
+
+        if obj is not None:
+            self._obj_to_name[obj] = candidate
+        return candidate
+
+    def _is_illegal_name(self, name: str, obj: Any) -> bool:
+        # 1. keywords are never allowed as names.
+        if name in keyword.kwlist:
+            return True
+
+        # 2. Can't shadow a builtin name, unless you *are* that builtin.
+        if name in builtins.__dict__:
+            return obj is not builtins.__dict__[name]
+
+        return False
+
+    def _is_used(self, name: str) -> bool:
+        return name in self._used_names
+
+    def _use_name(self, name: str):
+        self._used_names.setdefault(name)
+
+
+class _Globals:
+    """Tracks all external objects referenced from a Graph.
+
+    We assign each external object a unique global name, which is how it's
+    referenced in the Graph's generated Python code.
+
+    A dictionary of global name -> obj is passed to 'exec' to resolve
+    external references to their actual objects.
+    """
+    def __init__(self, graph_namer: _NameContext):
+        # The global name -> external object mapping.
+        self.globals: Dict[str, Any] = {}
```

Comment:
TODO: `_Globals` and `_NameContext` have a lot of overlap, I'm thinking currently about how to simplify that relationship.